### DISCUSSION TITLE
Debugging webhooks

### DISCUSSION THREAD

**Discussion by TannerGabriel** [argoproj-labs/argocd-notifications#334] [Q&A]
Hey, I'm trying to write a Webhook but am running into some trouble like the webhook not seeming to execute, but I haven't found a great way to debug webhook executions. Therefore I wanted to ask if there is a good way to find out the following things:

- Information if the webhook was executed
- Response code and information

**Comment by ryota-sakamoto** [comment]
You can get controller log to investigate triggerring webhook or not.

**Comment by TannerGabriel** [comment]
@ryota-sakamoto Thank you for the answer. I have looked at the logs and it seems like the triggers are alright, but I'm still not getting the webhook to my API endpoint:

```
time="2021-09-13T18:38:55Z" level=info msg="Start processing" app=argocd/podtatohead
time="2021-09-13T18:38:55Z" level=info msg="Trigger on-sync-running result: [{[0].xGlbhcF_taGYrI3UsrQsINn2hl0 [app-sync-running] false}]" app=argocd/podtatohead
time="2021-09-13T18:38:55Z" level=info msg="Trigger on-sync-succeeded result: [{[0].zxM90Et6k4Elb1-fHdjtDJq0xR0 [app-sync-succeeded] true}]" app=argocd/podtatohead
time="2021-09-13T18:38:55Z" level=info msg="Notification about condition 'on-sync-succeeded.[0].zxM90Et6k4Elb1-fHdjtDJq0xR0' already sent to '{keptn }'" app=argocd/podtatohead
time="2021-09-13T18:38:55Z" level=info msg="Processing completed" app=argocd/podtatohead
```

On the other hand, when I trigger the webhook manually using the following command I receive the webhook:

```
kubectl exec -n argocd -it argocd-notifications-controller--  /app/argocd-notifications template notify keptn-deployment-started podtatohead --recipient keptn:keptn-deployment-start
```

I also included my current triggers and application annotations for debugging purposes.

**Triggers:**

```
trigger.sync-operation-change: |
    - when: app.status.operationState.phase in ['Succeeded']
      send: [keptn-deployment-finished]
    - when: app.status.operationState.phase in ['Running']
      send: [keptn-deployment-started]
    - when: app.status.operationState.phase in ['Error', 'Failed']
      send: [keptn-deployment-finished]
```

**Application Annotations:**

```
notifications.argoproj.io/subscribe.on-sync-running.keptn
notifications.argoproj.io/subscribe.on-sync-succeeded.keptn
notified.notifications.argoproj.io= {"on-sync-succeeded:[0].zxM90Et6k4Elb1-fHdjtDJq0xR0:keptn:":1631550658}
```

The last annotation is autogenerated by Argo.

**Reply by wayttjl** [reply]
when i used webhook i set the debug level to 'debug' and got some error message.
maybe you can try it!
